File: 0399-evaluate-division/0399-evaluate-division.cpp
class Solution {
    map<string, set<string>> smap;
    map<pair<string, string>, double> dmap;
public:
    double dfs(double d, string start, string end, set<string>& visited) {
        //cout << start << "->" << end << ":" << d << " ";
        if (start==end) return d;
        for (string s:smap[start]) {
            if (!visited.count(s)) {
                visited.insert(s);
                pair<string, string> p={start, s};
                double ans=dfs(d*dmap[p], s, end, visited);
                if (ans!=-1.0) return ans;
                visited.erase(s);
            }
        }
        return -1.0;
    }
    
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        int n=values.size();
        for (int i=0; i<n; i++) {
            string s1=equations[i][0];
            string s2=equations[i][1];
            pair<string, string> p={s1, s2};
            dmap[p]=values[i];
            smap[s1].insert(s2);
            p={s2,s1};
            dmap[p]=1.0/values[i];
            smap[s2].insert(s1);
        }
        /*for (auto &it:dmap) {
            cout << it.first.first << "->" << it.first.second << ":" << it.second << endl;
        }
        for (auto &it:smap) {
            cout << it.first << ":";
            for (string s:it.second) cout << s << ",";
            cout << endl;
        }*/
        n=queries.size();
        vector<double> ans(n);
        for (int i=0; i<n; i++) {
            string s1=queries[i][0];
            string s2=queries[i][1];
            if (smap.find(s1)==smap.end() or smap.find(s2)==smap.end()) ans[i]=-1.0;
            else if (s1==s2) ans[i]=1.0;
            else if (smap[s1].count(s2)) {
                pair<string, string> p={s1, s2};
                ans[i]=dmap[p];
            } else {
                set<string> visited={s1};
                double d=1.0;
                ans[i]=dfs(1.0, s1, s2, visited);
            }
        }
        return ans;
    }
};

File: 0399-evaluate-division/0399-evaluate-division.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        A = defaultdict(lambda: {})
        for (a, b), val in zip(equations, values):
            A[a][b] = val
            A[b][a] = 1. / val
        print(A)
        ans = []
        for a, b in queries:
            if a not in A or b not in A:
                ans.append(-1)
                continue
            cur = [(a, 1.)]
            d = {a: 1.}
            done = False
            while cur:
                node, val = cur.pop()
                if node == b:
                    ans.append(val)
                    done = True
                    break
                for child, val2 in A[node].items():
                    if child in d:
                        continue
                    cur.append((child, val * val2))
                    d[child] = val * val2
            # print(d.items())
            if not done:
                ans.append(-1)
        return ans